packages/go_router/lib/src/misc/stateful_navigation_shell.dart

```diff
@@ -0,0 +1,185 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+import 'package:flutter/widgets.dart';
+
+import '../configuration.dart';
+import '../typedefs.dart';
+
+/// [InheritedWidget] for providing a reference to the closest
+/// [StatefulNavigationShellState].
+class InheritedStatefulNavigationShell extends InheritedWidget {
+  /// Constructs an [InheritedStatefulNavigationShell].
+  const InheritedStatefulNavigationShell({
+    required super.child,
+    required this.state,
+    super.key,
+  });
+
+  /// The [StatefulNavigationShellState] that is exposed by this InheritedWidget.
+  final StatefulNavigationShellState state;
+
+  /// The [StatefulShellRouteState] that is exposed by this InheritedWidget.
+  StatefulShellRouteState get routeState => state.routeState;
+
+  @override
+  bool updateShouldNotify(
+      covariant InheritedStatefulNavigationShell oldWidget) {
+    return state != oldWidget.state;
```

Comment:
This will not rebuild dependent if routeState has changed, i.e. for widget let depends on StatefulShellRoute.of(context).index.

Reply:
If this compares routeState, you may need to implement == operator in StatefulShellRouteState, otherwise, it will keep notify listen whenever StatefulNavigationShellState rebuilds because it will keep recreating StatefulShellRouteState.

Reply:
Yes, you may have a point there. I did't see the point of it having to be rebuilt when you're anyway rebuilding the `IndexedStack` when changing index. However since I've now opened up for custom navigator container implementations, the `InheritedStatefulNavigationShell` must probably anyway be rebuilt to prevent unexpected behaviour for the user. And rebuilding whenever the routeState changes would anyway be the most consistent and expected behaviour for `InheritedStatefulNavigationShell`. 

So I'll change `InheritedStatefulNavigationShell` to use `StatefulShellRouteState` (and implement ==).